include/aws/common/atomic.inl

```diff
@@ -0,0 +1,110 @@
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#if defined(__GNUC__) || defined(__clang__)
+
+    static inline int aws_atomic_load(int *dst) {
+        return __sync_val_compare_and_swap(dst, 0, 0);
+    }
+
+    static inline int aws_atomic_store(int *dst, int value) {
+        return __sync_lock_test_and_set(dst, value);
+    }
+
+    static inline int aws_atomic_add(int *dst, int addend) {
+        return __sync_fetch_and_add(dst, addend);
+    }
+
+    static inline bool aws_atomic_compare_exchange(int *dst, int *compare, int value) {
+        int comp = *compare;
+        int ret = __sync_val_compare_and_swap(dst, comp, value);
+        if (ret == comp) {
+            return true;
+        } else {
+            *compare = ret;
+            return false;
+        }
+    }
+
+    static inline void *aws_atomic_load_ptr(void **dst) {
+        return __sync_val_compare_and_swap(dst, 0, 0);
+    }
+
+    static inline void *aws_atomic_store_ptr(void **dst, void *value) {
+        return __sync_lock_test_and_set(dst, value);
+    }
+
+    static inline bool aws_atomic_compare_exchange_ptr(void **dst, void **compare, void *value) {
+        void *comp = *compare;
+        void *ret = __sync_val_compare_and_swap(dst, comp, value);
+        if (ret == comp) {
+            return true;
+        } else {
+            *compare = ret;
+            return false;
+        }
+    }
+
+#elif defined _WIN32
+
+    #include <intrin.h>
+
+    static inline int aws_atomic_load(int *dst) {
+        return _InterlockedCompareExchange((long*)dst, 0, 0);
```

Comment:
I addressed this idea earlier but wanted to also comment here: I think sticking with a small set of atomic operations, all of which are full memory barriers, would be the best choice after considering a variety of pros/cons. Though of course, it is still open for discussion!

If some pressing use-cases came justifying the need for more specific access patterns then I'd be happy to sink time into extending the API.

Reply:
The question of what sort of memory barrier is implied is independent of the level of cache-line access that is required. It's certainly possible to implement a full memory barrier with a shared cache line access. This could be done in x86 with e.g.:

```
MFENCE
MOV (%rax), %rax
MFENCE
```

Note that the actual load is not a barrier itself and could be reordered with adjacent instructions on either side; thus, we need to specify barriers on both sides if we want a "full memory barrier" on the operation.

It's worth looking at the generated assembler for this as well, by the way:

```
    bar = _InterlockedCompareExchange(&foo, 0, 0);
00007FF6F65616DC  xor         eax,eax  
00007FF6F65616DE  mov         dword ptr [rbp+0C0h],eax  
00007FF6F65616E4  lea         rcx,[foo (07FF6F656C180h)]  
00007FF6F65616EB  xor         eax,eax  
00007FF6F65616ED  mov         edx,dword ptr [rbp+0C0h]  
00007FF6F65616F3  lock cmpxchg dword ptr [rcx],edx  
00007FF6F65616F7  mov         dword ptr [bar (07FF6F656C184h)],eax  
```

You'll note that we did a simple load into eax, then did a `lock cmpxchg` (which asserts exclusive ownership of the specific cache line in question). Thus, this effectively acts as a read barrier after the actual read we care about (because locked instructions act as a memory barrier and have a total ordering); however, it does not prevent the read from being migrated earlier in the instruction stream.

That being said, we could accomplish the same thing with an on-stack locked operation, with less contention:

```
int aws_atomic_load(volatile int *dst) {
  // Note! The volatile int *dst _must_ be volatile to ensure this load is not
  // reordered by the compiler
  int value = *dst;
  volatile long *memory_barrier = 0;
  _InterlockedExchange(&memory_barrier, 0);
  return value;
}
```

We get the same memory barrier properties (on x86, at least), but we do not incur the penalty of acquiring exclusive access to the cache line.

This particular issue is going to be performance critical - only one processor can work on an exclusively-owned cache line at a time, while a shared cache line can be, well, shared. For atomic operations that do not intend to write to the cache line in question, this means that concurrency is reduced to, essentially, one thread.

It's also vital that we have the correct atomics semantics from the start, to ensure that any code written using them is not making incorrect assumptions. Changing things later on to use weaker barriers or shared access semantics may break code.

Reply:
> This particular issue is going to be performance critical ...

If this is the case, more functions can be added with more specific and fine-grained semantics. For now it seems we are going to stick with this small, tight, and simple API where each function issues a full memory barrier. We can adjust the documentation to make the memory barrier implication clear, though I don't see a problem simply adding more functions later when a real use-case makes itself evident. Later on more functions for weaker barriers or other semantics can be appended to the end of the `atomic.h` file, and their memory barrier implications can be explicitly documented as they are added over time.